hllplus/sparse.go

```diff
@@ -1 +1,256 @@
+	sparsePos, rho := computePosRhoW(hash, s.sparsePrecision)
+	delta := s.sparsePrecision - s.normalPrecision
+
+	// Check if the normal rhoW can be re-constructed from the lowest sp-p bits of the sparse
+	// index. In that case, we do not need to encode it explicitly.
+	if mask := uint32(1<<delta) - 1; sparsePos&mask != 0 {
+		return sparsePos
+	}
+
+	// Use the normal index instead of the sparse index since the lowest sp-p bits are all 0
+	// anyway (see the mask above).
+	normPos := sparsePos >> delta
+	return s.encodedFlag | normPos<<sparseRhoWBits | uint32(rho)
+}
+
+// --------------------------------------------------------------------
+
+type uint32Set map[uint32]struct{}
+
+func (s uint32Set) Add(n uint32) {
+	s[n] = struct{}{}
+}
+
+func (s uint32Set) Len() int {
+	return len(s)
+}
+
+func (s uint32Set) Clone() uint32Set {
+	if s == nil {
+		return nil
+	}
+
+	t := make(uint32Set, len(s))
+	for n := range s {
+		t.Add(n)
+	}
+	return t
+}
+
+func (s uint32Set) Flush() []uint32 {
+	nums := make(uint32Slice, 0, len(s))
+	for n := range s {
+		nums = append(nums, n)
+		delete(s, n)
+	}
+	sort.Sort(nums)
+	return nums
+}
+
+type uint32Slice []uint32
+
+func (p uint32Slice) Len() int           { return len(p) }
+func (p uint32Slice) Less(i, j int) bool { return p[i] < p[j] }
+func (p uint32Slice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+// --------------------------------------------------------------------
+
+// Varint encoded series of uint32s.
+type uvarintSlice []byte
+
+func (s uvarintSlice) Append(x uint32) uvarintSlice {
+	for x >= 0x80 {
+		s = append(s, byte(x)|0x80)
+		x >>= 7
+	}
+	return append(s, byte(x))
+}
+
+func (s uvarintSlice) Iterate(fn func(uint32)) {
+	t := s
+	for {
+		x, m := binary.Uvarint(t)
+		if m < 1 {
+			break
+		}
+		fn(uint32(x))
+		t = t[m:]
+	}
+}
+
+// --------------------------------------------------------------------
+
+var deltaSlicePool sync.Pool
+
+// Delta encoded slice of uint32s.
+type deltaSlice struct {
+	nums uvarintSlice
+	last uint32
+	size int
+}
+
+func recycleDeltaSlice(size int) *deltaSlice {
+	if v := deltaSlicePool.Get(); v != nil {
+		return v.(*deltaSlice)
+	}
+	return &deltaSlice{nums: make(uvarintSlice, 0, size)}
+}
+
+func (s *deltaSlice) Len() int {
+	return len(s.nums)
+}
+
+func (s *deltaSlice) Count() int {
+	return s.size
+}
+
+func (s *deltaSlice) Reset() {
+	s.nums = s.nums[:0]
+	s.last = 0
+	s.size = 0
+}
+
+func (s *deltaSlice) Release() {
+	s.Reset()
+	deltaSlicePool.Put(s)
+}
+
+func (s *deltaSlice) Clone() *deltaSlice {
+	if s == nil {
+		return nil
+	}
+
+	t := &deltaSlice{
+		nums: make(uvarintSlice, len(s.nums)),
+		last: s.last,
+		size: s.size,
+	}
+	copy(t.nums, s.nums)
+	return t
+}
+
+func (s *deltaSlice) Append(x uint32) {
+	s.nums = s.nums.Append(x - s.last)
+	s.last = x
+	s.size++
+}
+
+func (s *deltaSlice) Iterate(fn func(uint32)) {
+	var last uint32
+	s.nums.Iterate(func(u uint32) {
+		x := u + last
+		fn(x)
+		last = x
+	})
```

Comment:
@dim just confirming that this is right ^^

(any concerns about overflows?)

Reply:
No, not concern, the maximum value that can be stored in this "slice" is `(1 << p)` and our `p` can never be > 25 (so it definitely fits a uint32). The delta encoded slice is always sorted first so the values are always ascending